controllers/customer_controller.go

```diff
@@ -0,0 +1,203 @@
+	finalizerUtil "github.com/stakater/operator-utils/util/finalizer"
+	reconcilerUtil "github.com/stakater/operator-utils/util/reconciler"
+)
+
+const (
+	// TODO: Check if this is required in our case
+	// defaultRequeueTime = 60 * time.Second
+	CustomerFinalizer string = "jiraservicedesk.stakater.com/customer"
+)
+
+// CustomerReconciler reconciles a Customer object
+type CustomerReconciler struct {
+	client.Client
+	Log                   logr.Logger
+	Scheme                *runtime.Scheme
+	JiraServiceDeskClient jiraservicedeskclient.Client
+}
+
+// +kubebuilder:rbac:groups=jiraservicedesk.stakater.com,resources=customers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=jiraservicedesk.stakater.com,resources=customers/status,verbs=get;update;patch
+
+func (r *CustomerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	log := r.Log.WithValues("customer", req.NamespacedName)
+
+	log.Info("Reconciling Customer")
+
+	// Fetch the Customer instance
+	instance := &jiraservicedeskv1alpha1.Customer{}
+
+	err := r.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcilerUtil.DoNotRequeue()
+		}
+		// Error reading the object - donot requeue the request
+		return reconcilerUtil.RequeueWithError(err)
+	}
+
+	// Validate Custom Resource
+	if ok, err := instance.IsValid(); !ok {
+		return reconcilerUtil.ManageError(r.Client, instance, err, false)
+	}
+
+	// Resource is marked for deletion
+	if instance.DeletionTimestamp != nil {
+		log.Info("Deletion timestamp found for instance " + req.Name)
+		if finalizerUtil.HasFinalizer(instance, CustomerFinalizer) {
+			return r.handleDelete(req, instance)
+		}
+		// Finalizer doesn't exist so clean up is already done
+		return reconcilerUtil.DoNotRequeue()
+	}
+
+	// Add finalizer if it doesn't exist
+	if !finalizerUtil.HasFinalizer(instance, CustomerFinalizer) {
+		log.Info("Adding finalizer for instance" + req.Name)
+
+		finalizerUtil.AddFinalizer(instance, CustomerFinalizer)
+
+		err := r.Client.Update(context.TODO(), instance)
+		if err != nil {
+			return reconcilerUtil.ManageError(r.Client, instance, err, false)
+		}
+
+		return reconcilerUtil.DoNotRequeue()
+	}
+
+	return r.handleCreate(req, instance)
+}
+
+func (r *CustomerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&jiraservicedeskv1alpha1.Customer{}).
+		Complete(r)
+}
+
+func (r *CustomerReconciler) handleCreate(req ctrl.Request, instance *jiraservicedeskv1alpha1.Customer) (ctrl.Result, error) {
+	log := r.Log.WithValues("customer", req.NamespacedName)
+
+	if len(instance.Status.AccountId) == 0 {
```

Comment:
This needs to be checked in reconciler to determine if it's an event against update or create